pulsar-client/src/main/java/org/apache/pulsar/client/impl/AutoClusterFailover.java

```diff
@@ -0,0 +1,232 @@
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.client.api.AutoClusterFailoverBuilder;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.ServiceUrlProvider;
+
+@Slf4j
+@Data
+public class AutoClusterFailover implements ServiceUrlProvider {
+    private PulsarClient pulsarClient;
+    private volatile String currentPulsarServiceUrl;
+    private final String primary;
+    private final String secondary;
+    private final long failoverDelayNs;
+    private final long switchBackDelayNs;
+    private final ScheduledExecutorService executor;
+    private long recoverTimestamp;
+    private long failedTimestamp;
+    private final int interval = 30_000;
+    private final int TIMEOUT = 30_000;
+
+    private AutoClusterFailover(String primary, String secondary, long failoverDelayNs, long switchBackDelayNs) {
+        this.primary = primary;
+        this.secondary = secondary;
+        this.failoverDelayNs = failoverDelayNs;
+        this.switchBackDelayNs = switchBackDelayNs;
+        this.currentPulsarServiceUrl = primary;
+        this.recoverTimestamp = -1;
+        this.failedTimestamp = -1;
+        this.executor = Executors.newSingleThreadScheduledExecutor(
+                new DefaultThreadFactory("pulsar-service-provider"));
+    }
+
+    @Override
+    public void initialize(PulsarClient client) {
+        this.pulsarClient = client;
+
+        // start to probe primary cluster active or not
+        this.executor.scheduleAtFixedRate(catchingAndLoggingThrowables(() -> {
+            if (currentPulsarServiceUrl.equals(primary)) {
+                // current service url is primary, probe whether it is down
+                probeAndUpdateServiceUrl(secondary);
+            } else {
+                // current service url is secondary, probe whether it is down
+                probeAndUpdateServiceUrl(primary);
+                // secondary cluster is up, check whether need to switch back to primary
+                probeAndCheckSwitchBack(primary);
+            }
+        }), getInterval(), getInterval(), TimeUnit.MILLISECONDS);
+
+    }
+
+    @Override
+    public String getServiceUrl() {
+        return this.currentPulsarServiceUrl;
+    }
+
+    @Override
+    public void close() {
+        this.executor.shutdown();
+    }
+
+    boolean probeAvailable(String url) {
+        try {
+            String hostAndPort = parseHostAndPort(url);
+            if (Strings.isNullOrEmpty(hostAndPort)) {
+                return false;
+            }
+
+            Socket socket = new Socket();
+            socket.connect(new InetSocketAddress(parseHost(hostAndPort), parsePort(hostAndPort)), TIMEOUT);
```

Comment:
This will check for TCP connectivity, which is a good start, though there can be many cases in which it will give the false impression that cluster is healthy: 
 * We're connecting to a Pulsar proxy, but there are no available brokers
 * Using Istio on server side, which always accept the connection even if the broker is in a bad state
 *  We might have deadlocks in (all) brokers and while the connections get accepted, the brokers are not able to serve them.

We should consider to have a more in depth test to: 
 1. Check that we can authenticate with brokers
 2. (Maybe) Estimate how many brokers are up & healthy in the cluster

Reply:
Yes, the current probe port method has many disadvantages. We can use this method first, and in next step, we'd better provide a health check command, just like Zookeeper `ruok` command, to probe the healthy of the target Pulsar cluster.